### DISCUSSION TITLE
Capturing emulator output

### DISCUSSION THREAD

**Discussion by nk64** [nk64/GT4286#13] [General]
1. Pick an [emulator](https://github.com/nk64/GT4286/blob/main/Emulators.md) to study eg ```/emus/gamebatte/gambatte_sdl``` (note that there is a misspelling in the directory name of this emulator) (also note that ```/emus/gbc``` is unused)
2. Rename the emulator  by adding ```.bin``` to the filename eg. ```/emus/gamebatte/gambatte_sdl``` becomes ```/emus/gamebatte/gambatte_sdl.bin```
3. Then replace emulator with the shell script below.
4. After you play a game in that emulator you can find
    - the script output in eg. ```/emus/gamebatte/gambatte_sdl.log```
    - the emulator output in eg. ```/emus/gamebatte/gambatte_sdl.bin.log```

```
#!/bin/sh
logfile="$0.log"; log () { echo $1 >> $logfile && /bin/sync; }
wrapped_exe="$0.bin"
wrapped_exe_logfile="$wrapped_exe.log"
log_exe=true

log "---------------------------------------------------"
log "Arg0: $0"
log "ArgAll: $@"
log "Current Working Directory: $PWD"
log "Wrapped Exe: $wrapped_exe"
log "Executing wrapped exe: $wrapped_exe"

$wrapped_exe "$@" > $wrapped_exe_logfile 2>&1 && /bin/sync
```

**Comment by nk64** [comment]
I'm writing these instructions from memory so please let me know if you run into any difficulties.

**Reply by horsiezoo** [reply]
gotcha, thanks ! I get the gist now, I actually got logcat running earlier, very well done !

**Comment by nk64** [comment]
The emulators generate a ton of debug logging and every button press is recorded verbosely which _might_ cause excessive wear and tear on your SD Card.

**Reply by horsiezoo** [reply]
Thanks for this, for some reason the Filecopy script doesn’t dump anything to the designated directories 
Can confirm have renamed stock fbneo file to fbneo.old and renamed Filecopy.sh to fbneo and opened a MAME rom from games console.
I manually created sd …./roms/MAME….. /.* folders as well but nothing gets dumped 
I then edited the script to dump to sd:/output, same result.
Other scripts eg logcat, ls, meminfo all work, thanks again!
Eg: 

#!/bin/sh
logfile="/mnt/extsd/roms/MAME/output/log.txt"; log () { echo $1 >> $logfile && /bin/sync; }
log "FileCopy"

mkdir /mnt/extsd/roms/MAME/output/filecopy >> $logfile
mkdir /mnt/extsd/roms/MAME/output/filecopy/bin >> $logfile
mkdir /mnt/extsd/roms/MAME/output/filecopy/etc >> $logfile
mkdir /mnt/extsd/roms/MAME/output/filecopy/config >> $logfile
mkdir /mnt/extsd/roms/MAME/output/filecopy/lib >> $logfile
mkdir /mnt/extsd/roms/MAME/output/filecopy/overlay >> $logfile
mkdir /mnt/extsd/roms/MAME/output/filecopy/sbin >> $logfile
mkdir /mnt/extsd/roms/MAME/output/filecopy/tmp >> $logfile

cp -R /bin/* /mnt/extsd/roms/MAME/output/filecopy/bin >> $logfile
cp -R /etc/* /mnt/extsd/roms/MAME/output/filecopy/etc >> $logfile
cp -R /config/* /mnt/extsd/roms/MAME/output/filecopy/config >> $logfile
cp -R /data/* /mnt/extsd/roms/MAME/output/filecopy/data >> $logfile
cp -R /system/* /mnt/extsd/roms/MAME/output/filecopy/system >> $logfile
cp -R /lib/* /mnt/extsd/roms/MAME/output/filecopy/lib >> $logfile
cp -R /overlay/* /mnt/extsd/roms/MAME/output/filecopy/overlay >> $logfile
cp -R /sbin/* /mnt/extsd/roms/MAME/output/filecopy/sbin >> $logfile
cp -R /tmp/* /mnt/extsd/roms/MAME/output/filecopy/tmp >> $logfile

/bin/sync

**Reply by nk64** [reply]
@horsiezoo You probably already know, but those scripts in ```sdcard_tweaks/roms/MAME``` are not designed to be copied over the top of the mame emulator and run one by one.

Rather, they are designed to show up like roms in the MAME section of the game menu (and we replace the (```/emus/mame/fbneo```) with a 'script launcher script' (```sdcard_tweaks/emus/mame/fbneo```)

You can find more about doing it that way in the section on [Exploring the OS like a Boss](https://github.com/nk64/GT4286/blob/main/Exploring%20the%20OS.md) (not the hard way)

And you've made sure that a folder called ```/roms/MAME/output``` exists? (I really should make the script make that if it doesn't already exist)

I will double check the filecopy.sh script but is there anything in the log file (```/roms/MAME/output/log.txt```) about it?

Edit: I have tested your way and the Boss way and I get stuff showing up in ```/roms/MAME/output/filecopy```

**Reply by horsiezoo** [reply]
Well then... now it decides to run and dump fine. I'm going to say related to headache I had most of yesterday and lack of concentration.
Thanks again for the assist! 
Your scripts are fine. User error.

**Reply by nk64** [reply]
I have noticed that there is one script that might not run fine in it's current state (```fbneo-dat-files```) because it expects the emulator file to be called ```/emus/fbneo/fbneo.bin```. The ```.bin``` extension is part of another tweak and won't be there for a stock card, so I will update the script to fallback to ```/emus/fbneo/fbneo``` if ```/emus/fbneo/fbneo.bin``` doesn't exist.

**Reply by horsiezoo** [reply]
Does this mean something like this should in theory work to be able to pass arguments to exectuables? 

#!/bin/sh
cd  /mnt/extsd/emus/gamebatte
./gambatte_sdl /mnt/extsd/roms/GBC/ChromaticTetris.gbc

**Reply by nk64** [reply]
> I have noticed that there is one script that might not run fine in it's current state (`fbneo-dat-files`) because it expects the emulator file to be called `/emus/fbneo/fbneo.bin`. The `.bin` extension is part of another tweak and won't be there for a stock card, so I will update the script to fallback to `/emus/fbneo/fbneo` if `/emus/fbneo/fbneo.bin` doesn't exist.

Hmm, the ```emu-help-*``` scripts also require that emulator to has been renamed to ```*.bin```.

**Reply by horsiezoo** [reply]
Its fine as is, I made a copy of each emu in their respective folders, and named one as *.bin to allow .sh scripts to run and emulators too.

Comically the same as you said here 
The end result is that you must have two files:

/emus/fbneo/fbneo.bin (The real emulator)
/emus/fbneo/fbneo (The shell script, that does some manipulation of the rom parameter and then calls the real emulator)

Actually maybe I'll make a .bat file that automatically applies all of the mods so far to a stock image

**Comment by nk64** [comment]
Yes indeed, check out the [Emulators](https://github.com/nk64/GT4286/blob/main/Emulators.md) page for the reportedly available command line options (where I was able to extract them).

If you try any out, please let me know which ones work and which don't so I can update that page.

eg. the ```-menu``` option of fbneo seems to have no effect but as we know the ```-dat``` option does work.

The one I want to try but haven't got around to is for gambatte_sdl:
```-s, --scale N            Scale video output by an integer factor of N```

Update: ```--scale 1``` had no impact on the gameboy screen size :-(

**Reply by horsiezoo** [reply]
Yes I tried that myself yesterday -v (video filter works though) I tried -f (fullscreen) and -s 0.5, 0.75, and so on together to 0 effect. 
Am looking into gamedb db file to figure out if can change emulator names, eg: FC -> NES and rename MAME to SCRIPTS or the like

**Reply by nk64** [reply]
> Am looking into gamedb db file to figure out if can change emulator names, eg: FC -> NES and rename MAME to SCRIPTS or the like 

We can do this if we patch the ```/bin/gmenu2x```

**Reply by horsiezoo** [reply]
Hah yes I came to same conclusion, was poking around trying to figure out the offsets for those
